docker/tensorflow-application/video-serving/streaming.py

```diff
@@ -0,0 +1,61 @@
+from flask import Flask, Response, request
+from flask import render_template, send_file
+import threading
+import argparse
+import datetime
+import imutils
+import time
+import cv2
+import requests
+import numpy as np
+from PIL import Image
+from flask import jsonify
+import time
+
+
+outputFrame = None 
+lock = threading.Lock()
+
+app = Flask(__name__, template_folder='templates')
+
+@app.route("/")
+def index():
+	return render_template('index.html')
+
+def stream_to_memory():
+	global outputFrame, lock	
+	list1 = request.json['instances']
+	for l in list1:
+		img=np.asarray(l, dtype=np.uint32)
+		outputFrame = Image.fromarray(img.astype('uint8'))
```

Comment:
You have the the tensor flow analyzer sending images to the streamer, you put it in memory and then you have the streamer serve that image from another route? That actually makes me pretty nervous. When you're making a web application, one of it's primary benefits is that the HTTP protocol is stateless and your web application should be stateless as well. If there is some state (like in this case images) it should be external to the application: like in a database, filesystem, etc. A request should not be dependent on the in-memory state of the process serving it.

Not only will you be dealing with locks as things try to read/write to the same global (btw, don't use globals, they are almost always a bad idea, a "code smell"), but this application will never scale. And I'm not talking about scaling to a global audience, I'm talking about scaling past a single user. It might work for this demo, but is definitely something we'll need to fix longer term.

Reply:
@mpeters I agree with you. Additionally, after talking with @AyushAmbastha and understanding the application better, there are more inherent design issues with this application that will prevent it from scaling as well. For example, analysis.py and streaming.py are tightly coupled, where analysis.py handles 1 video stream at 1 HTTP route, and pushes the results to the 1 HTTP route exposed by streaming.py. This makes it fundamentally impossible to scale beyond 1 camera stream or even more than 1 replica. I've suggested to @AyushAmbastha that he add a TODO to look at using something like a kafka queue to decouple both applications, and in the immediate short term he could try to use a queue instead of a global variable to address some of the low FPS issues he was facing.